cvat/apps/annotation/annotation.py

```diff
@@ -77,13 +78,59 @@ def serialize(self):
         if serializer.is_valid(raise_exception=True):
             return serializer.data
 
+    @staticmethod
+    def _is_shape_inside(shape, start, stop):
+        return start <= int(shape['frame']) <= stop
+
+    @staticmethod
+    def _is_track_inside(track, start, stop):
+        # a <= b
+        def has_overlap(a, b):
+            return 0 <= min(b, stop) - max(a, start)
+
+        prev_shape = None
+        for shape in track['shapes']:
+            if prev_shape and not prev_shape['outside'] and \
+                has_overlap(prev_shape['frame'], shape['frame']):
+                    return True
+            prev_shape = shape
+
+        if not prev_shape['outside'] and prev_shape['frame'] <= stop:
+            return True
+
+        return False
+
+    @staticmethod
+    def _slice_track(track_, start, stop):
+        track = copy.deepcopy(track_)
+        segment_shapes = [s for s in track['shapes'] if AnnotationIR._is_shape_inside(s, start, stop)]
+
+        if len(segment_shapes) < len(track['shapes']):
+            interpolated_shapes = TrackManager.get_interpolated_shapes(track, start, stop)
+
+            for shape in interpolated_shapes:
```

Comment:
@azhavoro , could you please think how to reorganize the code below? It looks messy. 

It seems that get_interpolated_shapes doesn't respect start and stop by a reason. Thus let's filter them:

```
scoped_shapes = [s for s in interpolated_shapes['shapes'] if AnnotationIR._is_shape_inside(s, start, stop)]
if scoped_shapes and scoped_shapes[0]['outside']:
    scoped_shapes.pop(0) 
```

If scoped_shapes empty then nothing to do. Otherwise if `not scoped_shapes[0]['keyframe']`, let's add the shape into segment_shapes at the beginning. If `not scoped_shapes[-1]['keyframe']`, lets' add the shape into segment_shapes at the end.

Let's assign `track['interpolated_shapes'] = scoped_shapes`.

Of course I didn't check the logic. You need just improve your code and I hope the code above will give you some ideas how to do that.